docs/single-to-multi-master.md

```diff
@@ -17,9 +17,174 @@ not on the end user services. During this downtime, existing pods will continue
+```
+Example:
+
+```bash
+$ kops create ig master-us-east-1c --subnet us-east-1c --role Master
+```
+
+ * ``maxSize`` and ``minSize`` should be 1,
+ * only one zone should be listed.
+
+#### b - Create third master instance group
+
+Instance group for the third master, in a different AZ from the existing one, is
+also required. 
+
+```bash
+$ kops create instancegroup master-<availability-zone3> --subnet <availability-zone3> --role Master
+```
+
+Example:
+
+```bash
+$ kops create ig master-us-east-1d --subnet us-east-1d --role Master
+```
+
+ * ``maxSize`` and ``minSize`` should be 1,
+ * only one zone should be listed.
+
+### 4 - Update Cluster YAML to reference your new masters.
+
+#### a - Reference the new masters in your ectd configuration
 
-### a - Backup main etcd cluster
+
+```bash
+$ kops edit cluster example.com
+```
+
+ * In ``.spec.etcdClusters`` add 2 new members in each cluster, one for each new
+ availability zone.
+
+```yaml
+    - instanceGroup: master-<availability-zone2>
+      name: <availability-zone2-name>
+    - instanceGroup: master-<availability-zone3>
+      name: <availability-zone3-name>
+```
+
+Example:
+
+```yaml
+etcdClusters:
+  - etcdMembers:
+    - instanceGroup: master-eu-west-1a
+      name: a
+    - instanceGroup: master-eu-west-1b
+      name: b
+    - instanceGroup: master-eu-west-1c
+      name: c
+    name: main
+  - etcdMembers:
+    - instanceGroup: master-eu-west-1a
+      name: a
+    - instanceGroup: master-eu-west-1b
+      name: b
+    - instanceGroup: master-eu-west-1c
+      name: c
+    name: events
+```
+
+#### b - Update your cluster to reference the subnets for the new masters if it is not already configured.
+
+eg. 
+```
+subnets:
+  - cidr: 10.255.204.0/22
+    id: subnet-XXXXXX
+    name: sa-east-1a
+    type: Private
+    zone: us-east-1c
+  - cidr: 10.255.195.0/24
+    id: subnet-XXXXXX
+    name: utility-sa-east-1b
+    type: Utility
+    zone: us-east-1c
+  - cidr: 10.255.200.0/22
+    id: subnet-XXXXXX
+    name: sa-east-1a
+    type: Private
+    zone: us-east-1d
+  - cidr: 10.255.193.0/24
+    id: subnet-XXXXXX
+    name: utility-sa-east-1b
+    type: Utility
+    zone: us-east-1d
+```
+### 5 - Update Cluster to launch new masters
+
+```bash
+$ kops update cluster example.com --yes
+```
+
+### 6 - Issue a restore command for etcd-main and etcd-events
+
+```bash
+etcd-manager-ctl -backup-store s3://mystatestore/cluster.example.com/backups/etcd/main restore-backup 2019-06-12T10:13:40Z-000051
+
+etcd-manager-ctl -backup-store s3://mystatestore/cluster.example.com/backups/etcd/events restore-backup 2019-06-12T10:13:40Z-000051
+```
+
+  *Note that this does not start the restore immediately; you need to restart etcd on all masters 
+(or roll your masters quickly).*
+
+### - Force reboot your masters so etcd-manager will run the restore command you just issued.
+
+```bash
+kops rolling-update cluster cluster.example.com --yes \
+  --cloudonly
+  --force
+  --instance-group masters
```

Comment:
So... this whole process wasn't entirely successful for me. I ended up with 4 endpoints for the API service instead of 3, the endpoint from the old master was not removed when the machine was removed. The end result was 25% of new connections to the API via the internal service (ie, the kubernetes svc in the default namespace) were blackholed.

It looks like the 'lease' endpoint reconciler is not working (I'm using kops 1.12.2 and k8s 1.14.4). I ended up recovering the cluster by deleting the old lease (/registry/masterleases/<old-ip>) directly from etcd. Are watches backed up by etcd-manager? If so it should clear /registry/masterleases upon restore.

Reply:
1. Kops supports a `--instance-group-roles` flag. So the cli could be:
   ```
   kops rolling-update cluster cluster.example.com --yes \
     --cloudonly \
     --force \
     --instance-group-roles Master
   ```
2. I experienced exactly the same outcome as @drobinson123 described:   
   - `kubectl get ep -n default kubernetes -o yaml` shows an old master's IP  
   - Deleting the ip from `/registry/masterleases/` with etcdctl fixed the api-server errors   

After fixing the svc endpoint, my cluster's single-to-multi-master ops seems to succeed. I'm using kops `1.13.2`, k8s `v1.13.11` and etcd3 with etcd-manager.